packages/core/src/scp-cf/destination-accessor.ts

```diff
@@ -153,348 +168,420 @@ export async function getDestinationFromDestinationService(
-  name: string,
-  userJwt: string | DecodedJWT | undefined,
-  options: DestinationRetrievalOptions
-): Promise<Destination> {
-  const destinationServiceCreds = getDestinationServiceCredentials();
-  const accessToken = await serviceToken('destination', {
-    userJwt,
-    ...options
-  });
-
-  return fetchDestination(
-    destinationServiceCreds.uri,
-    accessToken,
-    name,
-    options
-  );
-}
+  private isProviderNeeded(): boolean {
+    if (this.selectionStrategy === alwaysSubscriber) {
+      return false;
+    }
 
-async function getDestinationWithAuthTokens(
-  name: string,
-  userJwt: string,
-  options?: DestinationRetrievalOptions
-): Promise<Destination> {
-  const destinationService = getService('destination');
-
-  if (!destinationService) {
-    throw Error(
-      `Failed to fetch destination "${name}"! No binding to a destination service found.`
-    );
+    return true;
   }
 
-  const accessToken = await userApprovedServiceToken(
-    userJwt,
-    destinationService,
-    options
-  );
-  return fetchDestination(
-    destinationService.credentials.uri,
-    accessToken,
-    name,
-    options
-  );
-}
+  private isSubscriberNeeded(
+    decodedUserJwt: DecodedJWT | undefined
+  ): decodedUserJwt is DecodedJWT {
+    if (!this.decodedUserJwt) {
+      return false;
+    }
 
-function getDestinationServiceCredentials(): DestinationServiceCredentials {
-  const credentials = getDestinationServiceCredentialsList();
-  if (!credentials || credentials.length === 0) {
-    throw Error(
-      'No binding to a Destination service instance found. Please bind a destination service instance to your application!'
-    );
+    if (this.selectionStrategy === alwaysProvider) {
+      return false;
+    }
+
+    if (this.isProviderAndSubscriberSameTenant()) {
+      return false;
+    }
+
+    return true;
   }
 
-  return credentials[0];
-}
+  private async searchProviderAccountForDestination(): Promise<
+    DestinationSearchResult | undefined
+  > {
+    if (!this.isProviderNeeded()) {
+      return;
+    }
 
-const emptyDestinationByType: DestinationsByType = {
-  instance: [],
-  subaccount: []
-};
+    if (this.useCache) {
+      return (
+        this.getProviderDestinationCache() ??
+        this.getProviderDestinationService()
+      );
+    }
 
-function subscriberDestinationIsSelected(
-  selectionStrategy: DestinationSelectionStrategy
-): boolean {
-  return (
-    selectionStrategy === subscriberFirst ||
-    selectionStrategy === alwaysSubscriber
-  );
-}
+    return this.getProviderDestinationService();
+  }
 
-function emptyDestinationsByType(
-  destinationByType: DestinationsByType
-): boolean {
-  return (
-    !destinationByType.instance.length && !destinationByType.instance.length
-  );
+  private async searchSubscriberAccountForDestination(): Promise<
+    DestinationSearchResult | undefined
+  > {
+    if (!this.isSubscriberNeeded(this.decodedUserJwt)) {
+      return;
+    }
```

Comment:
I think semantically it would fit better to move this up to the calling function.